keps/sig-cloud-provider/2699-add-webhook-hosting-to-ccm/README.md

```diff
@@ -0,0 +1,527 @@
+as two different binaries based on command line flags.
+
+#### Story 2
+
+For Cloud Providers who would like to keep things simple, it is easy to create
+a single process which handles both controllers and webhooks.
+
+#### Story 3
+
+PVL use case. Cloud Providers want to allow customers to migrate an existing
+workload to Kubernetes. That workload uses an existing persistent volume. To
+get that workload migrated the end user needs to be able to link the existing
+PV into the cluster. However this requires an association which requires calls
+out to the cloud provider for certain kinds of storage. Ideally the lookup and
+label of the PV to that pre-existing storage happens inline when the PV is 
+written. That ensures the write volume is attached to the Node/Pod when it is
+scheduled and there are no race conditions.
+
+### Notes/Constraints/Caveats (Optional)
+
+
+### Risks and Mitigations
+
+Potentially there could be problems running webhooks and controller in the same
+process. Delays of 10 seconds or more can cause webhooks to fail. It is important 
+to understand than irrespective of failure mode on the webhook coniguration, 
+timeouts will always turn a webhook call into a FAIL. As such we are making it
+easy to easily turn the CCM into two processses to mitigate this. It will be upto
+the Cloud Provider to determine if they want the webhook policy to be FAIL or 
+IGNORE. We will default to IGNORE as its the safe option. Incorrectly setting FAIL
+can quickly lead to a non functional cluster. Having a FAIL policy on Pods for 
+example can prevent the system from allocating the webhook service, which prevents
+the webhook from ever passing.
+
+Webhooks are configured by a runtime resource. As a consequence this configuration
+can be modified to deleted at runtime. That means that an admin on the cluster can
+disable or alter the functionality. This potentially makes it harder for a cloud 
+provider to enforce that this logic is being applied. It also means that there 
+needs to be a deployment mechanism for the webhook. It is left to the Cloud 
+Provider to determine if the need for an inline request is sufficient to override
+these concerns. The Cloud Provider can alternately use the controller route which 
+is not inline or use an actual admission controller, built into the APIServer.
+
+We are actually changing the framework which generates the CCM and not the CCM
+itself. It has been pointed out that it is not the role of the controller
+manager to run webhooks. Controller managers should run controllers and webhooks
+are not controllers. As we are modifying the framework, we should consider this
+as we can create two processes. The CCM which just has controllers in it. We
+can also create a Cloud Webhook Manager. That is being left as homework for the 
+Cloud Provider. However the sample CCM which demonstrates how this will be done
+will have both in the same sample to make it easy.
```

Comment:
Is it worth mentioning that there is prior art here? Controller runtime runs controllers and webhooks from the same manager, so there are many existing projects out there following this pattern already.

Cluster API for example runs all the webhooks and controllers in the same binaries under the same manager. We could discuss with communities who are familiar with this about problems they've had to understand what problems CCMs might have with this pattern too